DirectProgramming/DPC++FPGA/ReferenceDesigns/cholesky/src/cholesky_demo.cpp

```diff
@@ -0,0 +1,351 @@
+#include <math.h>
+
+#include <CL/sycl.hpp>
+#include <list>
+#include <sycl/ext/intel/ac_types/ac_complex.hpp>
+#include <sycl/ext/intel/fpga_extensions.hpp>
+
+// dpc_common.hpp can be found in the dev-utilities include folder.
+// e.g., $ONEAPI_ROOT/dev-utilities//include/dpc_common.hpp
+#include "cholesky.hpp"
+#include "dpc_common.hpp"
+
+// #define DEBUG
+
+/*
+  COMPLEX, COLS_COMPONENT, ROWS_COMPONENT and FIXED_ITERATIONS are defined
+  by the build system.
+  Depending on the value of COMPLEX, the real or complex Cholesky decomposition
+  is defined (A = LL*)
+
+  Function arguments:
+  - a_matrix:    The input matrix.
+  - l_matrix     The L matrix. The function will overwrite this matrix.
+                 The vector will only contain the lower triangular elements
+                 of the matrix, in a row by row fashion.
+  - q:           The device queue.
+  - matrix_count: Number of matrices to decompose.
+  - repetitions: The number of repetitions of the computation to execute.
+                 (for performance evaluation)
+*/
+#if COMPLEX == 0
+// Real single precision floating-point Cholesky decomposition
+void CholeskyDecomposition(std::vector<float> &a_matrix,
+                           std::vector<float> &l_matrix, sycl::queue &q,
+                           int matrix_count, int repetitions) {
+  constexpr bool is_complex = false;
+  CholeskyDecompositionImpl<COLS_COMPONENT, ROWS_COMPONENT, FIXED_ITERATIONS,
+                            is_complex, float>(a_matrix, l_matrix, q,
+                                               matrix_count, repetitions);
+}
+#else
+// Complex single precision floating-point Cholesky Decomposition
+void CholeskyDecomposition(std::vector<ac_complex<float> > &a_matrix,
+                           std::vector<ac_complex<float> > &l_matrix,
+                           sycl::queue &q, int matrix_count, int repetitions) {
+  constexpr bool is_complex = true;
+  CholeskyDecompositionImpl<COLS_COMPONENT, ROWS_COMPONENT, FIXED_ITERATIONS,
+                            is_complex, float>(a_matrix, l_matrix, q,
+                                               matrix_count, repetitions);
+}
+#endif
+
+/*
+  Returns true if both the real and complex parts of the given ac_complex
+  value are finite
+*/
+bool IsFinite(ac_complex<float> val) {
+  return std::isfinite(val.r()) && std::isfinite(val.i());
+}
+
+/*
+  Returns true if the given value is finite
+*/
+bool IsFinite(float val) { return std::isfinite(val); }
+
+/*
+  Returns a random floating-point value between min and max
+*/
+float RandomValueInInterval(float min, float max) {
+  return min + static_cast<float>(rand()) /
+                   (static_cast<float>(RAND_MAX) / (max - min));
+}
```

Comment:
Do you really care about randomness here? If so, you could use some of the C++ random classes. However, this is probably fine.

Reply:
Yeah there is probably a more modern way of doing so, I'm just used to doing it the old C way because managing the distribution of the randomness of the inputs is not really necessary and this is just to test that the kernel is working ok